puppy/ui/replpane.py

```diff
@@ -0,0 +1,99 @@
+from PyQt5.QtWidgets import QTextEdit
+from PyQt5.QtGui import QTextCursor
+from PyQt5.QtCore import QIODevice
+from PyQt5.QtCore import Qt
+from PyQt5.QtSerialPort import QSerialPort, QSerialPortInfo
+
+# TODO:
+#   - shutdown serial port cleanly on exit
+#   - use monospace font
+#   - get backspace and arrow keys working
+
+MICROBIT_PID = 516
+MICROBIT_VID = 3368
+
+
+def find_microbit():
+    """
+    Returns the port for the first microbit it finds connected to the host
+    computer. If no microbit is found, returns None.
+    """
+    available_ports = QSerialPortInfo.availablePorts()
+    for port in available_ports:
+        pid = port.productIdentifier()
+        vid = port.vendorIdentifier()
+        if pid == MICROBIT_PID and vid == MICROBIT_VID:
+            return port.portName()
+    return None
+
+
+class REPLPane(QTextEdit):
+    """
+    REPL = Read, Evaluate, Print, Loop.
+
+    This widget represents a REPL client connected to a BBC micro:bit.
+    """
+    def __init__(self, port, parent=None):
+        super().__init__(parent)
+        self.setAcceptRichText(False)
+        self.setReadOnly(False)
+        self.setLineWrapMode(QTextEdit.NoWrap)
+        self.setObjectName('replpane')
+        # A flag to indicate that we've sent some sort of escape
+        # character e.g. \b
+        self.escape_flag = False
+
+        # open the serial port
+        self.serial = QSerialPort(self)
+        self.serial.setPortName(port)
+        self.serial.setBaudRate(115200)
+        print(self.serial.open(QIODevice.ReadWrite))
+        self.serial.readyRead.connect(self.on_serial_read)
+
+        # clear the text
+        self.clear()
+
+    def on_serial_read(self):
+        self.append(self.serial.readAll())
+
+    def keyPressEvent(self, data):
+        self.escape_flag = False
+        text = data.text()
+        msg = bytes(text, 'utf8')
+        key = data.key()
+        if key == Qt.Key_Backspace:
+            msg = '\b'
+            self.delete()
+            self.escape_flag = True
+        elif key == Qt.Key_Up:
+            msg = '\x1B[A'
+        elif key == Qt.Key_Down:
+            msg = '\x1B[B'
+        self.serial.write(msg)
+
+    def append(self, data):
+        txt = str(data, 'utf8')
+        if self.escape_flag:
+            return
+        tc = self.textCursor()
+        tc.movePosition(QTextCursor.End)
+        self.setTextCursor(tc)
+        self.insertPlainText(txt)
+        self.ensureCursorVisible()
+
+    def delete(self):
```

Comment:
Absolutely, it's a hack.

I did have a solution that worked similarly to how you suggest but it quickly got very complicated when dealing with all the junk coming back from the micro:bit (I presume terminal codes of some sort = b'\x01[K', b'\x01[22' and so on - off the top of my head...). 

Unfortunately, such codes could be split over several calls to `on_serial_read` so I built a shonky FSM to keep track of things. This quickly got out of hand and I decided to back out and revert to the simpler but hacky version you see before you. It was very brittle and I realised I didn't know what I was doing.

Ultimately, I wanted something that made it semi-useful with a view to making it work properly once I've got my head around the problem and looked into re-using potential existing solutions (I'm guessing there's code in pyserial for exactly this sort of case that we can re-use or at the very least study in order to write our own solution).

Thoughts..?